experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+  unsigned bitwidth;
+  std::vector<bool> value;
+
+public:
+  PortChannel ctrl, outs;
+
+  Constant(unsigned tempBitwidth, const std::vector<bool> &tempValue)
+      : bitwidth(tempBitwidth), value(tempValue) {}
+  void exec(bool clk = true) override {
+    outs.data->updateValue(value);
+    outs.valid = ctrl.valid;
+    ctrl.ready = outs.ready;
+  }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    ctrl.valid->printValue();
+    llvm::outs() << "\n";
+    ctrl.ready->printValue();
+    llvm::outs() << "\n";
+    outs.valid->printValue();
+    llvm::outs() << "\n";
+    outs.ready->printValue();
+    llvm::outs() << "\n";
+    outs.data->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "Constant"; }
+  ~Constant(){};
+};
+class DelayBuffer : public Component {};
+
+class End : public Component {
+  unsigned bitwidth;
+
+public:
+  PortChannel ins, outs;
+  End(unsigned tempBitwidth) : bitwidth(tempBitwidth) {}
+  void exec(bool clk = true) override {
+    outs = ins;
+    outs.valid = ins.valid;
+    conjunction(ins.valid, outs.ready, ins.ready);
+  }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    ins.valid->printValue();
+    llvm::outs() << "\n";
+    ins.ready->printValue();
+    llvm::outs() << "\n";
+    ins.data->printValue();
+    llvm::outs() << "\n";
+    outs.valid->printValue();
+    llvm::outs() << "\n";
+    outs.ready->printValue();
+    llvm::outs() << "\n";
+    outs.data->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "End"; }
+  ~End() {}
+};
+
+class Extsi : public Component {
+  unsigned inputWidth;
+  unsigned outputWidth;
+
+public:
+  PortChannel ins, outs;
+  Extsi(unsigned tempInputWidth, unsigned tempOutputWidth)
+      : inputWidth(tempInputWidth), outputWidth(tempOutputWidth) {}
+  void exec(bool clk = true) override {
+    if (ins.data->hasValue()) {
+      // little endian
+      auto inVec = std::any_cast<std::vector<bool>>(ins.data->getData());
+      std::vector<bool> outVec(outputWidth, inVec.front());
+      for (unsigned i = outputWidth - inputWidth; i < outputWidth; ++i) {
+        outVec[i] = inVec[i + inputWidth - outputWidth];
+      }
+      outs.data->updateValue(outVec);
+    }
+    outs.valid = ins.valid;
+    ins.ready = outs.ready;
+  }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    ins.valid->printValue();
+    llvm::outs() << "\n";
+    ins.ready->printValue();
+    llvm::outs() << "\n";
+    ins.data->printValue();
+    llvm::outs() << "\n";
+    outs.valid->printValue();
+    llvm::outs() << "\n";
+    outs.ready->printValue();
+    llvm::outs() << "\n";
+    outs.data->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "Extsi"; }
+  ~Extsi() {}
+};
+
+class Join : public Component {};
+class Muli : public Component {
+  unsigned bitwidth;
+
+public:
+  PortChannel lhs, rhs, result;
```

Comment:
As we discussed before, this information is already part of the operation's interface and should not be part of the execution model itself.